src/modules/schema-validator/schema-validator.service.ts

```diff
@@ -0,0 +1,76 @@
+import { JSONSchema7 } from 'json-schema'
+import { CeramicClient } from '@ceramicnetwork/http-client'
+import { TileDocument } from '@ceramicnetwork/stream-tile'
+import { IndexedDocument } from '../../types/indexedDocument.type'
+import { NotImplementedException } from '../../exceptions/not-implemented.exception'
+
+export class SchemaValidator {
+  ceramic: CeramicClient
+  constructor(ceramic: CeramicClient) {
+    this.ceramic = ceramic
+  }
+
+  /**
+   * Registers schema with ceramic
+   */
+  public async registerSchema(schema: JSONSchema7): Promise<TileDocument> {
+    // TODO - figure out if we need to capture and index the ceramic ID of the schema
+
+    const created = await TileDocument.create(
+      this.ceramic,
+      {
+        $schema: 'http://json-schema.org/draft-07/schema#',
+        title: 'QuickStartSchema',
+        type: 'object',
+        properties: {
+          id: { type: 'string' },
+          foo: { type: 'string' },
+          revision: { type: 'number' },
+        },
+        required: ['foo'],
+      },
+      {
+        controllers: [this.ceramic?.did?.id || ''],
+        family: 'schema',
+      },
+    )
+
+    return created
+  }
+
+  /**
+   * Deregisters schema with ceramic
+   */
+  public async deregisterSchema(schema: JSONSchema7) {
+    // TODO - figure out if we can deregister a schema based on the schema definition itself or if we need the ceramic document ID
+    throw new NotImplementedException('deregisterSchema')
+  }
+
+  /**
+   * Crawls indexed subgraph and validates schema of each document in the subgraph vs the advertised schema
+   */
+  public async validateSubgraphVsSchema(subgraphRoot: IndexedDocument, schema: JSONSchema7) {
```

Comment:
So for one the `indexedDocument` is going to be an array. The subgraph is an array of manifests of each subgraph document.